sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala

```diff
@@ -384,239 +384,166 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
-      condition = Some(attribute === attribute)),
-    OutputMode.Append(),
-    Seq("without a watermark in the join keys"))
+      // stream-stream allowed with range condition yielding state value watermark
+      assertSupportedInStreamingPlan(
+        s"$joinType join with stream-stream relations and state value watermark",
+        leftRelation.join(rightRelation, joinType = joinType,
+          condition = Some(attribute > timeWithWatermark + 10)),
+        OutputMode.Append())
 
-  // Left semi joins: stream-stream allowed with range condition yielding state value watermark
-  assertSupportedInStreamingPlan(
-    "left semi join with stream-stream relations and state value watermark", {
-      val leftRelation = streamRelation
-      val rightTimeWithWatermark =
-        AttributeReference("b", IntegerType)().withMetadata(watermarkMetadata)
-      val rightRelation = new TestStreamingRelation(rightTimeWithWatermark)
-      leftRelation.join(
-        rightRelation,
-        joinType = LeftSemi,
-        condition = Some(attribute > rightTimeWithWatermark + 10))
-    },
-    OutputMode.Append())
-
-  // Left semi joins: stream-stream not allowed with insufficient range condition
-  assertNotSupportedInStreamingPlan(
-    "left semi join with stream-stream relations and state value watermark", {
-      val leftRelation = streamRelation
-      val rightTimeWithWatermark =
-        AttributeReference("b", IntegerType)().withMetadata(watermarkMetadata)
-      val rightRelation = new TestStreamingRelation(rightTimeWithWatermark)
-      leftRelation.join(
-        rightRelation,
-        joinType = LeftSemi,
-        condition = Some(attribute < rightTimeWithWatermark + 10))
-    },
-    OutputMode.Append(),
-    Seq("appropriate range condition"))
-
-  // Left anti joins: stream-* not allowed
-  testBinaryOperationInStreamingPlan(
-    "left anti join",
-    _.join(_, joinType = LeftAnti),
-    streamStreamSupported = false,
-    batchStreamSupported = false,
-    expectedMsg = "Left anti join")
+      // stream-stream not allowed with insufficient range condition
+      assertNotSupportedInStreamingPlan(
+        s"$joinType join with stream-stream relations and state value watermark",
+        leftRelation.join(rightRelation, joinType = joinType,
+          condition = Some(attribute < timeWithWatermark + 10)),
+        OutputMode.Append(),
+        Seq("is not supported without a watermark in the join keys, or a watermark on " +
+          "the nullable side and an appropriate range condition"))
+    }
 
-  // Right outer joins: stream-* not allowed
-  testBinaryOperationInStreamingPlan(
-    "right outer join",
-    _.join(_, joinType = RightOuter),
-    streamBatchSupported = false,
-    streamStreamSupported = false,
-    expectedMsg = "outer join")
+    // stream-stream inner join doesn't emit late rows, whereas outer joins could
+    Seq((Inner, false), (LeftOuter, true), (RightOuter, true)).map {
```

Comment:
Note: this test is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L744 .